l3kernel/l3str-convert.dtx

```diff
@@ -2567,6 +2580,102 @@
 \group_end:
 %    \end{macrocode}
 %
+% \subsection{PDF names and strings by expansion}
+%
+% \begin{macro}[EXP]{\str_convert_pdfname:n, \@@_convert_pdfname:n}
+% \begin{macro}[EXP]{\@@_convert_pdfname_first:w, \@@_convert_pdfname_loop:w}
+% \begin{macro}[EXP]{\@@_convert_pdfname_outer:n}
+% \begin{macro}[EXP]{\@@_convert_pdfname_inner:w}
+% \begin{macro}[EXP]
+%   {\@@_convert_pdfname_bytes:n, \@@_convert_pdfname_bytes_aux:n}
+% \begin{macro}[EXP]{\@@_convert_pdfname_bytes_aux:nnn}
+%   To convert to PDF names by expansion, we work purely on UTF-8 input. The
+%   first step is to make a string, after which we use a space-based mapping to
+%   divide up the input, then a simple token-by-token approach. In Unicode
+%   engines, we break down everything before one-byte codepoints, but for
+%   $8$-bit engines there is no need to worry. Actual escaping is covered
+%   by the same code as used in the non-expandable route.
+%    \begin{macrocode}
+\cs_new:Npn \str_convert_pdfname:n #1
+  {
+    \exp_args:Ne \tl_to_str:n
+      {
+         \exp_after:wN \@@_convert_pdfname:n \exp_after:wN
+           { \tl_to_str:n {#1} }
+      }
+  }
+\cs_new:Npx \@@_convert_pdfname:n #1
+  {
+    \exp_not:N \@@_convert_pdfname_first:w #1 ~
+      \exp_not:N \q_recursion_tail \c_space_tl
+      \exp_not:N \q_recursion_stop
+  }
+\cs_new:Npn \@@_convert_pdfname_first:w #1 ~
+  {
+    \quark_if_recursion_tail_stop:n {#1}
+    \@@_convert_pdfname_outer:n {#1}
+  }
+\cs_new:Npn \@@_convert_pdfname_loop:w #1 ~
+  {
+    \quark_if_recursion_tail_stop:n {#1}
+    \c_hash_str 20
+    \@@_convert_pdfname_outer:n {#1}
+  }
+\cs_new:Npx \@@_convert_pdfname_outer:n #1
+  {
+    \exp_not:N \@@_convert_pdfname_inner:w #1
+      \exp_not:N \q_recursion_tail \exp_not:N \q_recursion_stop
+    \exp_not:N \@@_convert_pdfname_loop:w
+  }
+\bool_lazy_or:nnTF
+  { \sys_if_engine_luatex_p: }
+  { \sys_if_engine_xetex_p: }
+  {
+    \cs_new:Npn \@@_convert_pdfname_inner:w #1
+      {
+        \quark_if_recursion_tail_stop:n {#1}
+        \int_compare:nNnTF { `#1 } > { "7F }
+          { \@@_convert_pdfname_bytes:n {#1} }
+          { \@@_escape_name_char:N #1 }
+        \@@_convert_pdfname_inner:w
+      }
+    \cs_new:Npn \@@_convert_pdfname_bytes:n #1
+      {
+        \exp_args:Ne \@@_convert_pdfname_bytes_aux:n
+          { \char_to_utfviii_bytes:n {`#1} }
```

Comment:
This line reminds me that we probably have utf8 conversion code both as this function `\char_to_utfviii_bytes:n` that you used, and internally in `\str_set_convert:Nnnn`.  Maybe worth merging.

Reply:
Possibly: `\char_to_utfviii_bytes:n` remains experimental, partly due to the name, partly as I'm still not 100% sure about the form of the output